mkosi/cage.py

```diff
@@ -0,0 +1,667 @@
+
+        if r < 0:
+            oserror(src)
+
+        try:
+            libc.move_mount.argtypes = (ctypes.c_int, ctypes.c_char_p, ctypes.c_int, ctypes.c_char_p, ctypes.c_uint)
+            r = libc.move_mount(fd, b"", AT_FDCWD, dst.encode(), MOVE_MOUNT_F_EMPTY_PATH)
+        except AttributeError:
+            libc.syscall.argtypes = (
+                ctypes.c_long, ctypes.c_int, ctypes.c_char_p, ctypes.c_int, ctypes.c_char_p, ctypes.c_uint)
+            r = libc.syscall(NR_move_mount, fd, b"", AT_FDCWD, dst.encode(), MOVE_MOUNT_F_EMPTY_PATH)
+
+        if r < 0:
+            oserror(dst)
+    finally:
+        os.close(fd)
+
+
+class umask:
+    def __init__(self, mask: int):
+        self.mask = mask
+
+    def __enter__(self) -> None:
+        self.mask = os.umask(self.mask)
+
+    def __exit__(self, *args: object, **kwargs: object) -> None:
+        os.umask(self.mask)
+
+
+def become_user(uid: int, gid: int) -> None:
+    """
+    This function implements the required dance to unshare a user namespace and map the current user to itself or to
+    root within it. The kernel only allows a process running outside of the unshared user namespace to write the
+    necessary uid and gid mappings, so we fork off a child process, make it wait until the parent process has unshared
+    a user namespace, and then writes the necessary uid and gid mappings.
+    """
+    ppid = os.getpid()
+
+    event = libc.eventfd(0, 0)
+    if event < 0:
+        oserror()
+
+    pid = os.fork()
+    if pid == 0:
+        try:
+            os.read(event, ctypes.sizeof(ctypes.c_uint64))
+            os.close(event)
+            with open(f"/proc/{ppid}/setgroups", "w") as f:
+                f.write("deny\n")
+            with open(f"/proc/{ppid}/gid_map", "w") as f:
+                f.write(f"{gid} {os.getgid()} 1\n")
+            with open(f"/proc/{ppid}/uid_map", "w") as f:
+                f.write(f"{uid} {os.getuid()} 1\n")
+        except OSError as e:
+            os._exit(e.errno)
+        except BaseException:
+            os._exit(1)
+        else:
+            os._exit(0)
+
+    try:
+        unshare(CLONE_NEWUSER)
+    finally:
+        os.write(event, ctypes.c_uint64(1))
+        os.close(event)
+        _, status = os.waitpid(pid, 0)
+
+    rc = os.waitstatus_to_exitcode(status)
+    if rc != 0:
+        raise OSError(rc, os.strerror(rc))
+
+
+def acquire_privileges(*, become_root: bool = False) -> bool:
+    if os.getuid() == 0 or (not become_root and have_effective_cap(CAP_SYS_ADMIN)):
+        return False
+
+    become_user(0 if become_root else os.getuid(), 0 if become_root else os.getgid())
+
+    if not become_root:
+        cap_permitted_to_ambient()
+
+    return True
+
+
+def userns_has_single_user() -> bool:
+    if not os.path.exists("/proc/self/uid_map"):
+        return False
+
+    with open("/proc/self/uid_map") as f:
+        lines = f.readlines()
+
+    return len(lines) == 1 and int(lines[0].split()[-1]) == 1
```

Comment:
Safes one check in the usual case of the file existing
```suggestion
    try:
        with open("/proc/self/uid_map") as f:
            lines = f.readlines()

        return len(lines) == 1 and int(lines[0].split()[-1]) == 1
    except FileNotFoundError:
        return False
```